src/profiler.ts

```diff
@@ -13,57 +13,222 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
+import * as http from 'http';
 import * as path from 'path';
+import * as pify from 'pify';
+import * as zlib from 'zlib';
 
-import {AuthenticationConfig, Common, Logger, Service, ServiceConfig} from '../third_party/types/common-types';
+import {perftools} from '../src/profile';
+import {AuthenticationConfig, Common, Logger, Service, ServiceConfig, ServiceObject, ServiceObjectConfig} from '../third_party/types/common-types';
 
-import {Config} from './config';
+import {ProfilerConfig} from './config';
 import {HeapProfiler} from './profilers/heap-profiler';
 import {TimeProfiler} from './profilers/time-profiler';
 
+export const common: Common = require('@google-cloud/common');
 const pjson = require('../../package.json');
-const common: Common = require('@google-cloud/common');
-
-export interface ProfilerConfig extends AuthenticationConfig {
-  projectId: string;
-  logLevel: number;
-  serviceContext: {service: string; version?: string;};
-  instance: string;
-  zone: string;
-  disableTime: boolean;
-  disableHeap: boolean;
+const API = 'https://cloudprofiler.googleapis.com/v2';
+const gzip = pify(zlib.gzip);
+
+const WALL_TYPE = 'WALL';
+const HEAP_TYPE = 'HEAP';
+
+/**
+ * Interface for body of response from Stackdriver Profiler API when creating
+ * profile and used as body of request to Stackdriver Profiler API when
+ * uploading a profile.
+ *
+ * Public for testing.
+ */
+export interface RequestProfile {
```

Comment:
This definition is basically equivalent to this: https://github.com/googleapis/googleapis/blob/master/google/devtools/cloudprofiler/v2/profiler.proto#L87.

But since we're using REST rather than gRPC, we aren't currently using profiler.proto.

Reply:
I think it's worth considering, but it does have a few imports. So (I think) to be able to compile it to javascript, we would need to add more than  just `profiler.proto`.

Reply:
It is okay to consider this out of scope for this PR, but I would still like to see this happen. Do you have the full list of protos we need? Presumably gRPC based profiler agents already have this full list in their repos?

Reply:
I agree we should do this, I'll comment on #27 with some details.